### PR TITLE
Feature/characters database

### PR DESCRIPTION
**Pull request by Sheroziy-CODE**
Implemented Database use case for Characters Fragment, in addition added Pull to refresh feature

### CODE REVIEW

File: app/src/main/java/rick_and_morty/data/model/RealmCharacterResultsDto.kt [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+open class RealmCharacters (
+    @PrimaryKey
+    var created: String = "",
+    var episode: RealmList<String> = RealmList(),
+    var gender: String = "",
+    var id: Int = 0,
+    var image: String = "",
+    var locationDto: RealmLocation? = null,
+    var name: String = "",
+    var originDto: RealmOrigin? = null,
```
Comment (kraffmann) [inline app/src/main/java/rick_and_morty/data/model/RealmCharacterResultsDto.kt]: 1) Please do not use implementation details in the naming -> `RealmLocation`, `Realm...`. If we would switch databases we would have to rename all the classes, too.

2) Can `RealmLocation` and `RealmOrigin` really be null? Is it possible that the API is not providing information for these fields?

File: app/src/main/java/rick_and_morty/data/realm/RealmInstance.kt [hunk 2]
```diff
@@ -1,10 +1,13 @@
 package rick_and_morty.data.realm
 
 import io.realm.RealmObject
+import rick_and_morty.data.model.CharacterResultsDto
 import rick_and_morty.data.model.episodes.EpisodeResultDto
 
 interface RealmInstance {
```
Comment (kraffmann) [inline app/src/main/java/rick_and_morty/data/realm/RealmInstance.kt]: I know I was suggesting the name, but maybe renaming it to something like `DatabaseInstance` or `LocalStorageInstance` would be better. Same for the implementation class.

File: app/src/main/java/rick_and_morty/ui/characters/CharactersViewModel.kt [hunk 3]
```diff
@@ -59,6 +59,33 @@ class CharactersViewModel @Inject constructor(
...
+    fun refreshCharacters() {
+        viewModelScope.launch {
+            try {
+                _characters.update { it.copy(isLoading = true) }
+
+                characterRepository.clearCharactersDatabase()
+                page = 1
+
+                val dbCharacters = characterRepository.getCharacters(page)
+                characterRepository.saveCharactersToDatabase(dbCharacters)
```
Comment (kraffmann) [inline app/src/main/java/rick_and_morty/ui/characters/CharactersViewModel.kt]: Please do not clear any data before you got a more up-to-date version of the data.
You might end up showing nothing.

This principle should also be applied to the `EpisodesViewModel` and the deletion of the database at app start.

File: app/src/test/java/rick_and_morty/rules/FakeRealmProvier.kt [hunk 4]
```diff
@@ -1,36 +1,60 @@
...
 import io.realm.RealmObject
+import rick_and_morty.data.model.CharacterResultsDto
+import rick_and_morty.data.model.RealmCharacters
 import rick_and_morty.data.model.episodes.EpisodeResultDto
 import rick_and_morty.data.model.episodes.realm.RealmEpisodes
 import rick_and_morty.data.realm.RealmInstance
+import rick_and_morty.ui.characters.CharactersMapper.toRealmCharacter
 import rick_and_morty.ui.episodes.EpisodesMapper.toRealmEpisode
 
 class FakeRealmInstance : RealmInstance {
```
Comment (kraffmann) [inline app/src/test/java/rick_and_morty/rules/FakeRealmProvier.kt]: For some reason the file name was not changed, when we went for `FakeRealmInstance`.

Comment (Sheroziy-CODE) [reply app/src/test/java/rick_and_morty/rules/FakeRealmProvier.kt]: Done

File: app/src/main/java/rick_and_morty/data/realm/RealmInstance.kt (hunk 2 above)
Comment (Sheroziy-CODE) [reply app/src/main/java/rick_and_morty/data/realm/RealmInstance.kt]: Done

File: app/src/main/java/rick_and_morty/data/model/RealmCharacterResultsDto.kt (hunk 1 above)
Comment (Sheroziy-CODE) [reply app/src/main/java/rick_and_morty/data/model/RealmCharacterResultsDto.kt]: 1. Done.

2. Not really, but it is data class for realm, and Realm does not support directly enforcing non-null constraints on object references.

File: app/src/main/java/rick_and_morty/ui/characters/CharactersViewModel.kt (hunk 3 above)
Comment (Sheroziy-CODE) [reply app/src/main/java/rick_and_morty/ui/characters/CharactersViewModel.kt]: Done

File: app/src/test/java/rick_and_morty/data/repository/EpisodesRepositoryTest.kt [hunk 5]
```diff
@@ -62,7 +62,7 @@ class EpisodeRepositoryTest {
     @Test
     fun `verify getEpisodesFromDatabase has been called`() = runBlocking {
 
-        val mockRealmEpisode = RealmEpisodes().apply {
+        val mockRealmEpisode = LocalStorageEpisodes().apply {
```
Comment (kraffmann) [inline app/src/test/java/rick_and_morty/data/repository/EpisodesRepositoryTest.kt]: NIT: Here is still a trace of `Realm`.

File: app/src/test/java/rick_and_morty/rules/FakeLocalStorageProvier.kt [hunk 6]
```diff
@@ -2,37 +2,37 @@ package rick_and_morty.rules
...
-    private val fakeCharactersDatabase = mutableListOf<RealmCharacters>()
+    private val fakeEpisodesDatabase = mutableListOf<LocalStorageEpisodes>()
+    private val fakeCharactersDatabase = mutableListOf<LocalStorageCharacters>()
 
     var isFindAllCalled = false
     private var isSaveEpisodesToDatabaseCalled = false
     private var isSaveCharactersToDatabaseCalled = false
 
-    fun setRealmEpisodeResults(realmEpisodeResults: List<RealmEpisodes>) {
+    fun setLocalStorageEpisodeResults(realmEpisodeResults: List<LocalStorageEpisodes>) {
```
Comment (kraffmann) [inline app/src/test/java/rick_and_morty/rules/FakeLocalStorageProvier.kt]: NIT: Realm ... Realm ... Realm ...

**Review by kraffmann — APPROVED** [review]

**Comment by kraffmann** [comment]
Was `FakeLocalStorageProvier.kt` renamend?

**Comment by Sheroziy-CODE** [comment]
> Was `FakeLocalStorageProvier.kt` renamend?

Yep it was

File: app/src/test/java/rick_and_morty/data/repository/EpisodesRepositoryTest.kt (hunk 5 above)
Comment (Sheroziy-CODE) [reply app/src/test/java/rick_and_morty/data/repository/EpisodesRepositoryTest.kt]: Done

File: app/src/test/java/rick_and_morty/rules/FakeLocalStorageProvier.kt (hunk 6 above)
Comment (Sheroziy-CODE) [reply app/src/test/java/rick_and_morty/rules/FakeLocalStorageProvier.kt]: Done

**Review by kraffmann — APPROVED** [review]

**Pull request merged**